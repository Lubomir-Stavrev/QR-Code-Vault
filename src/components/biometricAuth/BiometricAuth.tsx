import React, {useState, FC} from 'react';
import {Text, View, TouchableOpacity, ActivityIndicator} from 'react-native';

import {useQuery} from 'react-query';
import TouchID from 'react-native-touch-id';
import styles from '../../styles/AuthStyles';
interface Props {
  handlePinCodeSignIn: () => void;
  onSuccesfullAuthentication: () => void;
}

const BiometricAuth: FC<Props> = props => {
  const [errorMessage, setErrorMessage] = useState<string | null>();

  const optionalConfigObject = {
    title: 'Authentication Required',
    imageColor: '#FFCC1D',
    imageErrorColor: '#ff0000',
    sensorDescription: 'Touch sensor',
    cancelText: 'Cancel',
    fallbackLabel: 'Show Passcode',
    unifiedErrors: true,
    passcodeFallback: true,
    sensorErrorDescription: 'Too many attempts',
  };
  const {isLoading, isError, error} = useQuery<Boolean, Error>('authData', () =>
    TouchID.authenticate('Authenticate', optionalConfigObject).then(() =>
      props.onSuccesfullAuthentication(),
    ),
  );

  if (isError) {
    if (error.message !== 'User canceled authentication') {
      setErrorMessage('Something went worng');
    }
  }

  return (
    <View style={[styles.authContainer, styles.horizontal]}>
      {isLoading ? (
        <ActivityIndicator size="large" />
      ) : isError ? (
        <View>
          <Text style={styles.errorMessage}>{errorMessage}</Text>
          <TouchableOpacity onPress={() => props.handlePinCodeSignIn()}>
            <View style={styles.button}>
              <Text style={styles.signInButton}>Sign In with pin code</Text>
            </View>
          </TouchableOpacity>
        </View>
      ) : null}
    </View>
  );
};

export default BiometricAuth;
