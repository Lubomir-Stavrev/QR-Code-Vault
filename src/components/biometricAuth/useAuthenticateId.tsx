import {useMutation} from 'react-query';
import TouchID from 'react-native-touch-id';
interface Props {
  navigation: {navigate: (text: string) => void};
}
export const useAuthenticateId = (props: Props) => {
  const optionalConfigObject = {
    title: 'Authentication Required',
    imageColor: '#FFCC1D',
    imageErrorColor: '#ff0000',
    sensorDescription: 'Touch sensor',
    cancelText: 'Cancel',
    fallbackLabel: 'Show Passcode',
    unifiedErrors: true,
    passcodeFallback: true,
    sensorErrorDescription: 'Too many attempts',
  };

  const {
    mutate: authenticate,
    isLoading,
    isError,
  } = useMutation<Boolean, Error>(
    TouchID.authenticate('Authenticate', optionalConfigObject),
    {
      onSuccess: () => {
        props.navigation.navigate('QRCodeMenu');
      },
    },
  );

  return {authenticate, isLoading, isError};
};
