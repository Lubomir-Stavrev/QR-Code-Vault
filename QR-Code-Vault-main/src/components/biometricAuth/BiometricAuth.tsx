import React, {FC} from 'react';
import {Text, View, TouchableOpacity, ActivityIndicator} from 'react-native';

import {useAuthenticateId} from './useAuthenticateId';
import styles from '../../styles/AuthStyles';
interface Props {
  navigation: {navigate: (text: string) => void};
}

const BiometricAuth: FC<Props> = props => {
  const {isLoading, isError} = useAuthenticateId(props);

  return (
    <View style={[styles.authContainer, styles.horizontal]}>
      {isLoading ? (
        <ActivityIndicator testID={'ActivityIndicator'} size="large" />
      ) : isError ? (
        <View>
          <Text testID={'textUserCanceledAuth'} style={styles.errorMessage}>
            User canceled authentication
          </Text>
          <TouchableOpacity
            accessibilityLabel="pinCodeBackUpTouchableOpacity"
            onPress={() => props.navigation.navigate('PinCodeBackupAuth')}>
            <View style={styles.button}>
              <Text style={styles.signInButton}>Sign In with pin code</Text>
            </View>
          </TouchableOpacity>
        </View>
      ) : null}
    </View>
  );
};

export default BiometricAuth;
