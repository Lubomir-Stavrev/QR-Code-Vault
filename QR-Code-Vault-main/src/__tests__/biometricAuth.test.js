import React from 'react';
import BiometricAuth from '../components/biometricAuth/BiometricAuth';
import { useAuthenticateId } from '../components/biometricAuth/useAuthenticateId';
import { render, fireEvent } from '@testing-library/react-native';

jest.mock('../components/biometricAuth/useAuthenticateId', () => ({
    useAuthenticateId: jest.fn(),
}));

jest.mock('react-query', () => ({
    useQueryClient: jest.fn(),
}));

describe('does Biometric Authentication handles useAuthenticateId results correctly', () => {
    it('should display Activity indecator when isLoading is true', async() => {
        useAuthenticateId.mockImplementation(() => ({
            isLoading: true,
        }));
        const rendered = render( < BiometricAuth / > );
        const activityIndicatorComponent =
            rendered.getByTestId('ActivityIndicator');

        expect(activityIndicatorComponent).toBeTruthy();
    });
    it('should display text that the user canceled authentication when an error occures', async() => {
        useAuthenticateId.mockImplementation(() => ({
            isError: true,
        }));
        const errorMessage = 'User canceled authentication';
        const rendered = render( < BiometricAuth / > );
        const textComponent = rendered.getByTestId('textUserCanceledAuth');

        expect(textComponent.props.children).toEqual(errorMessage);
    });
});

describe('does Bimetric Authentication handles press events as it shoulds', () => {
            it('should calls navigate to PinCodeBackupAuth on press', () => {
                    const navigation = { navigate: jest.fn() };
                    const spy = jest.spyOn(navigation, 'navigate');
                    useAuthenticateId.mockImplementation(() => ({
                        isError: true,
                    }));
                    const rendered = render( < BiometricAuth navigation = { navigation }
                        />);
                        const pinCodeBackUpTouchableOpacityComponent = rendered.getByTestId(
                            'pinCodeBackUpTouchableOpacity',
                        ); fireEvent.press(pinCodeBackUpTouchableOpacityComponent); expect(spy).toBeCalledWith('PinCodeBackupAuth');
                    });
            });