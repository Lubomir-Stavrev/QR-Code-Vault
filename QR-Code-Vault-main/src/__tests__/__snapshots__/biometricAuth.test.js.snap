// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`are QR Code componets rendering successfully renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#202020",
      "flex": 1,
      "height": "100%",
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "height": 1334,
          "width": 750,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "white",
            "fontFamily": "Roboto",
            "fontSize": 35,
            "letterSpacing": 1,
          }
        }
      >
        QR Code Vault
      </Text>
      <View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#01a3ea",
                "borderRadius": 5,
                "justifyContent": "center",
                "margin": 10,
                "padding": 10,
                "paddingBottom": 15,
                "paddingTop": 15,
                "textAlign": "center",
                "width": "100%",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffff",
                  "fontFamily": "Roboto",
                  "fontSize": 22,
                  "fontWeight": "700",
                  "letterSpacing": 1,
                }
              }
            >
              Scan QR CODE
            </Text>
          </View>
        </View>
        <View
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#01a3ea",
                "borderRadius": 5,
                "justifyContent": "center",
                "margin": 10,
                "padding": 10,
                "paddingBottom": 15,
                "paddingTop": 15,
                "textAlign": "center",
                "width": "100%",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffff",
                  "fontFamily": "Roboto",
                  "fontSize": 22,
                  "fontWeight": "700",
                  "letterSpacing": 1,
                }
              }
            >
              View QR code collection
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`are QR Code componets rendering successfully renders correctly 2`] = `
Array [
  <Text
    style={
      Object {
        "fontSize": 32,
      }
    }
  >
    QR Codes Collection
  </Text>,
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "flex": 1,
        "paddingTop": null,
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#ffff",
          "borderRadius": 10,
          "left": 0,
          "opacity": 1,
          "padding": 10,
          "position": "absolute",
        }
      }
    >
      <View>
        <Text>
          Go Back
        </Text>
      </View>
    </View>
    <Text>
      Collection is empty
    </Text>
  </RCTSafeAreaView>,
]
`;

exports[`does Biometric Authentication renders successfully renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#202020",
        "flex": 1,
        "height": "100%",
        "top": "5%",
      },
      Object {
        "flexDirection": "row",
        "justifyContent": "space-around",
        "padding": 10,
      },
    ]
  }
/>
`;
